File: worker/src/worker/processor.py
"""Processing pipeline coordinating preprocessing, embeddings, clustering, LLM, and spike detection."""
from __future__ import annotations

import logging
import re
import time

import numpy as np

from .config import get_settings
from .embeddings import InstrumentedEmbeddingAdapter, get_embedding_adapter
from .clustering import Clusterer, ClusteringOutput
from .llm_adapter import InstrumentedLLMAdapter, get_llm_adapter
from .logger import get_logger, log_with_context
from .metrics import (
    worker_preprocessing_time_seconds,
)
from .spike_detector import SpikeDetector
from .types import Chunk, ChunkMetrics, ChunkResult, ClusterResult, Mention

logger = get_logger(__name__)


CLEAN_URL_RE = re.compile(r"https?://\S+")
CLEAN_WHITESPACE_RE = re.compile(r"\s+")


class ChunkProcessor:
    """Main processing pipeline for a chunk of mentions."""

    def __init__(
        self,
        worker_id: str,
        redis_client,
    ) -> None:
        self._settings = get_settings()
        self._worker_id = worker_id
        self._embedding_adapter: InstrumentedEmbeddingAdapter = get_embedding_adapter(worker_id)
        self._llm_adapter: InstrumentedLLMAdapter = get_llm_adapter(worker_id)
        self._clusterer = Clusterer(worker_id)
        self._spike_detector = SpikeDetector(redis_client, worker_id)

    async def process_chunk(self, chunk: Chunk, *, fetch_time_ms: float) -> ChunkResult:
        metrics = ChunkMetrics(io_time_ms=fetch_time_ms)
        total_start = time.perf_counter()

        with self._llm_adapter.context(brand=chunk.brand, chunk_id=chunk.chunk_id):
            mentions = self._preprocess(chunk, metrics)
            if not mentions:
                processing_ms = (time.perf_counter() - total_start) * 1000
                metrics.total_task_time_ms = processing_ms + metrics.io_time_ms
                return ChunkResult(
                    chunk_id=chunk.chunk_id,
                    brand=chunk.brand,
                    timestamp=int(chunk.created_at.timestamp()),
                    clusters=[],
                    metrics=metrics,
                )

            embeddings, embed_duration = await self._generate_embeddings(chunk, mentions)
            metrics.embedding_time_ms = embed_duration

            clustering_output = await self._perform_clustering(chunk, embeddings)
            metrics.clustering_time_ms = clustering_output.duration_ms

            clusters = await self._analyze_clusters(chunk, mentions, clustering_output, metrics)

        processing_ms = (time.perf_counter() - total_start) * 1000
        metrics.total_task_time_ms = processing_ms + metrics.io_time_ms
        log_with_context(
            logger,
            level=logging.INFO,
            message="Chunk processed",
            context={
                "worker_id": self._worker_id,
                "brand": chunk.brand,
                "chunk_id": chunk.chunk_id,
                "mentions": len(mentions),
                "clusters": len(clusters),
            },
            metrics=metrics.model_dump(),
        )

        return ChunkResult(
            chunk_id=chunk.chunk_id,
            brand=chunk.brand,
            timestamp=int(chunk.created_at.timestamp()),
            clusters=clusters,
            metrics=metrics,
        )

    def _preprocess(self, chunk: Chunk, metrics: ChunkMetrics) -> list[Mention]:
        start = time.perf_counter()
        dedup: dict[str, Mention] = {}
        for mention in chunk.mentions:
            cleaned = self._clean_text(mention.text)
            if not cleaned:
                continue
            if cleaned in dedup:
                continue
            dedup[cleaned] = Mention(
                id=mention.id,
                source=mention.source,
                text=cleaned,
                created_at=mention.created_at,
                sentiment=mention.sentiment,
                metadata=mention.metadata,
            )
        duration = time.perf_counter() - start
        metrics.preprocessing_time_ms = duration * 1000
        worker_preprocessing_time_seconds.labels(self._worker_id, chunk.brand).observe(duration)
        log_with_context(
            logger,
            level=logging.INFO,
            message="Preprocessing completed",
            context={
                "worker_id": self._worker_id,
                "brand": chunk.brand,
                "chunk_id": chunk.chunk_id,
                "original_mentions": len(chunk.mentions),
                "clean_mentions": len(dedup),
            },
            metrics={"preprocessing_time_ms": metrics.preprocessing_time_ms},
        )
        return list(dedup.values())

    async def _generate_embeddings(self, chunk: Chunk, mentions: list[Mention]) -> tuple[np.ndarray, float]:
        start = time.perf_counter()
        embeddings = await self._embedding_adapter.embed(
            [m.text for m in mentions],
            brand=chunk.brand,
            chunk_id=chunk.chunk_id,
        )
        duration = (time.perf_counter() - start) * 1000
        return embeddings, duration

    async def _perform_clustering(self, chunk: Chunk, embeddings: np.ndarray) -> ClusteringOutput:
        return await self._clusterer.cluster(embeddings, brand=chunk.brand, chunk_id=chunk.chunk_id)

    async def _analyze_clusters(
        self,
        chunk: Chunk,
        mentions: list[Mention],
        clustering_output: ClusteringOutput,
        metrics: ChunkMetrics,
    ) -> list[ClusterResult]:
        brand = chunk.brand
        chunk_id = chunk.chunk_id
        clusters: list[ClusterResult] = []
        llm_total_ms = 0.0
        spike_total_ms = 0.0

        for grouping in clustering_output.clusters:
            cluster_mentions = [mentions[idx] for idx in grouping.indices]
            texts = [mention.text for mention in cluster_mentions]
            examples = [mention.text for mention in cluster_mentions[: self._settings.preprocessing_examples]]

            cluster_start = time.perf_counter()
            summary = await self._llm_adapter.summarize(texts)
            sentiment = await self._llm_adapter.sentiment(texts)
            llm_total_ms += (time.perf_counter() - cluster_start) * 1000

            spike_start = time.perf_counter()
            spike_result = await self._spike_detector.detect(brand, grouping.cluster_id, len(cluster_mentions))
            spike_total_ms += (time.perf_counter() - spike_start) * 1000

            clusters.append(
                ClusterResult(
                    cluster_id=grouping.cluster_id,
                    count=len(cluster_mentions),
                    examples=examples,
                    summary=summary,
                    spike=spike_result.is_spike,
                    sentiment=sentiment,
                )
            )

        metrics.llm_time_ms = llm_total_ms
        metrics.spike_detection_time_ms = spike_total_ms
        return clusters

    @staticmethod
    def _clean_text(text: str) -> str:
        text = CLEAN_URL_RE.sub("", text)
        text = CLEAN_WHITESPACE_RE.sub(" ", text)
        text = text.strip().lower()
        return text

File: worker-rs/src/processor.rs
use std::collections::{HashMap, HashSet};
use std::sync::Arc;
use std::time::Instant;

use anyhow::Result;
use once_cell::sync::Lazy;
use regex::Regex;
use tracing::warn;

use crate::clustering::{Clusterer, ClusteringOutput};
use crate::config::Settings;
use crate::embeddings::InstrumentedEmbeddingAdapter;
use crate::llm::InstrumentedLlmAdapter;
use crate::metrics::{
    WORKER_PREPROCESSING_TIME_SECONDS,
};
use crate::spike::{SpikeDetectionResult, SpikeDetector};
use crate::types::{Chunk, ChunkMetrics, ChunkResult, ClusterResult, Mention};

static URL_RE: Lazy<Regex> = Lazy::new(|| Regex::new(r"https?://\S+").expect("Invalid URL regex"));
static WHITESPACE_RE: Lazy<Regex> = Lazy::new(|| Regex::new(r"\s+").expect("Invalid whitespace regex"));
const TOPIC_LIMIT: usize = 10;

pub struct Processor {
    settings: Arc<Settings>,
    embeddings: InstrumentedEmbeddingAdapter,
    clusterer: Clusterer,
    llm: InstrumentedLlmAdapter,
    spike_detector: SpikeDetector,
}

impl Processor {
    pub fn new(
        settings: Arc<Settings>,
        embeddings: InstrumentedEmbeddingAdapter,
        clusterer: Clusterer,
        llm: InstrumentedLlmAdapter,
        spike_detector: SpikeDetector,
    ) -> Self {
        Self {
            settings,
            embeddings,
            clusterer,
            llm,
            spike_detector,
        }
    }

    pub async fn process_chunk(&self, chunk: Chunk, fallback_brand: &str, fetch_time_ms: f64) -> Result<ChunkResult> {
        let total_start = Instant::now();
        let mut metrics = ChunkMetrics {
            io_time_ms: fetch_time_ms,
            ..Default::default()
        };

        let mut brand = chunk.brand.clone();
        if brand.trim().is_empty() {
            brand = fallback_brand.to_string();
        }

        let preprocess_start = Instant::now();
        let mentions = self.preprocess(&chunk.mentions);
        let preprocessing_duration = preprocess_start.elapsed();
        metrics.preprocessing_time_ms = preprocessing_duration.as_secs_f64() * 1000.0;
        WORKER_PREPROCESSING_TIME_SECONDS
            .with_label_values(&[&self.settings.worker_id, &brand])
            .observe(preprocessing_duration.as_secs_f64());

        if mentions.is_empty() {
            metrics.total_task_time_ms = total_start.elapsed().as_secs_f64() * 1000.0 + metrics.io_time_ms;
            return Ok(ChunkResult {
                chunk_id: chunk.chunk_id,
                brand,
                timestamp: chunk.created_at.timestamp(),
                clusters: Vec::new(),
                metrics,
            });
        }

        let embed_start = Instant::now();
        let embeddings = self
            .embeddings
            .embed(&mentions, &brand, &chunk.chunk_id)
            .await;
        metrics.embedding_time_ms = embed_start.elapsed().as_secs_f64() * 1000.0;

        let clustering_output = self
            .clusterer
            .cluster(&embeddings, &brand, &chunk.chunk_id)
            .await;
        metrics.clustering_time_ms = clustering_output.duration_ms;

        let clusters = self
            .build_cluster_results(&brand, &chunk.chunk_id, &mentions, clustering_output)
            .await;

        metrics.llm_time_ms = clusters.iter().map(|cluster| cluster.metrics.llm_ms).sum();
        metrics.spike_detection_time_ms = clusters
            .iter()
            .map(|cluster| cluster.metrics.spike_ms)
            .sum();

        let cluster_results: Vec<ClusterResult> = clusters.into_iter().map(|wrapper| wrapper.cluster).collect();

        metrics.total_task_time_ms = total_start.elapsed().as_secs_f64() * 1000.0 + metrics.io_time_ms;

        Ok(ChunkResult {
            chunk_id: chunk.chunk_id,
            brand,
            timestamp: chunk.created_at.timestamp(),
            clusters: cluster_results,
            metrics,
        })
    }

    fn preprocess(&self, mentions: &[Mention]) -> Vec<String> {
        let mut seen = HashSet::new();
        let mut cleaned = Vec::new();

        for mention in mentions {
            let candidate = self.clean_text(&mention.text);
            if candidate.is_empty() {
                continue;
            }
            if seen.insert(candidate.clone()) {
                cleaned.push(candidate);
            }
        }

        cleaned
    }

    fn clean_text(&self, text: &str) -> String {
        let without_urls = URL_RE.replace_all(text, "");
        let normalized = WHITESPACE_RE.replace_all(without_urls.trim(), " ");
        normalized.to_lowercase()
    }

    async fn build_cluster_results(
        &self,
        brand: &str,
        chunk_id: &str,
        mentions: &[String],
        clustering_output: ClusteringOutput,
    ) -> Vec<ClusterWithMetrics> {
        let mut results = Vec::new();

        for group in clustering_output.clusters {
            let cluster_mentions: Vec<String> = group
                .indices
                .iter()
                .filter_map(|&idx| mentions.get(idx).cloned())
                .collect();

            if cluster_mentions.is_empty() {
                continue;
            }

            let examples = cluster_mentions
                .iter()
                .take(self.settings.preprocessing_examples)
                .cloned()
                .collect::<Vec<_>>();

            let llm_start = Instant::now();
            let summary = self.llm.summarize(brand, &cluster_mentions).await;
            let sentiment = self.llm.sentiment(brand, &cluster_mentions).await;
            let llm_duration_ms = llm_start.elapsed().as_secs_f64() * 1000.0;

            let spike_start = Instant::now();
            let spike_result = match self
                .spike_detector
                .detect(brand, group.cluster_id, cluster_mentions.len())
                .await
            {
                Ok(result) => result,
                Err(err) => {
                    warn!(
                        worker_id = %self.settings.worker_id,
                        brand,
                        chunk_id,
                        cluster_id = group.cluster_id,
                        error = %err,
                        "Spike detection failed; marking cluster as non-spike"
                    );
                    SpikeDetectionResult::default()
                }
            };
            let spike_duration_ms = spike_start.elapsed().as_secs_f64() * 1000.0;

            let topics = cluster_mentions
                .iter()
                .take(TOPIC_LIMIT)
                .cloned()
                .collect::<Vec<_>>();

            results.push(ClusterWithMetrics {
                cluster: ClusterResult {
                    cluster_id: group.cluster_id,
                    count: cluster_mentions.len(),
                    examples: examples.clone(),
                    summary,
                    spike: spike_result.is_spike,
                    sentiment,
                    topics: Some(topics),
                },
                metrics: ClusterStageMetrics {
                    llm_ms: llm_duration_ms,
                    spike_ms: spike_duration_ms,
                },
            });
        }

        if results.is_empty() {
            // Fallback: treat all mentions as a single cluster.
            let examples = mentions
                .iter()
                .take(self.settings.preprocessing_examples)
                .cloned()
                .collect::<Vec<_>>();
            results.push(ClusterWithMetrics {
                cluster: ClusterResult {
                    cluster_id: 1,
                    count: mentions.len(),
                    examples: examples.clone(),
                    summary: examples.first().cloned(),
                    spike: false,
                    sentiment: HashMap::from([
                        ("positive".to_string(), 0.33),
                        ("negative".to_string(), 0.33),
                        ("neutral".to_string(), 0.34),
                    ]),
                    topics: Some(examples),
                },
                metrics: ClusterStageMetrics::default(),
            });
        }

        results
    }
}

#[derive(Default)]
struct ClusterStageMetrics {
    llm_ms: f64,
    spike_ms: f64,
}

struct ClusterWithMetrics {
    cluster: ClusterResult,
    metrics: ClusterStageMetrics,
}
